### PR TITLE
add validation to LedgerIndex

### PR DESCRIPTION
**Pull request by tekbar11**
Proposed fix for #93 .

Added a method `isValid` to LedgerIndex to check the validity of the wrapped index.

The `LedgerIndex.of` factory method now throws a `NumberFormatException` if the given value is not a valid index.

### CODE REVIEW

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java [hunk 1]
```diff
@@ -34,9 +34,17 @@ public LedgerIndex(String value) {
    * @param value A {@link String} containing either an integer or a shortcut.
    *
    * @return A {@link LedgerIndex} with the given value.
+   *
+   * @throws NumberFormatException if value is an invalid index
    */
-  public static LedgerIndex of(String value) {
-    return new LedgerIndex(value);
+  public static LedgerIndex of(String value)
```
Comment (sappenin) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:40]: Let's add some unit test coverage for the three text-values, and probably for a `null` value too?

**Review by sappenin — CHANGES_REQUESTED** [review]

**Comment by tekbar11** [comment]
Marked constructor in `LedgerIndex` as deprecated, with a javadoc comment explaining the reason of the deprecation.

Added unit tests to `LedgerIndexTest` for `LedgerIndex`'s factory methods.

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java (hunk 1 above)
Comment (tekbar11) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:46]: I added some unit tests as requested.

I'd also like to point out that the `LedgeIndexTest.createInvalidNumericalLedgerIndex` could be removed because it does not make sense to test the `LedgerIndex.plus` method after bad calls to `LedgerIndex.of`.

**Review by nkramer44 — APPROVED** [review]
Looks good to me! Thanks @tekbar11 for doing this!

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java [hunk 2]
```diff
@@ -45,8 +64,13 @@ public static LedgerIndex of(String value) {
    * @param value An {@link UnsignedLong} specifying a ledger index.
    *
    * @return A {@link LedgerIndex} with the given value as a {@link String}.
+   *
+   * @throws NullPointerException if value is null
    */
   public static LedgerIndex of(UnsignedLong value) {
+    if (value == null) {
```
Comment (nkramer44) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:71]: You could replace this with `Object.requireNonNull(value)`, which may make Codecov happy

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java [hunk 3]
```diff
@@ -34,9 +40,22 @@ public LedgerIndex(String value) {
...
+   *
+   * @throws NullPointerException if value is null
+   *
+   * @throws NumberFormatException if value is an invalid index
    */
-  public static LedgerIndex of(String value) {
-    return new LedgerIndex(value);
+  public static LedgerIndex of(String value)
+  throws NumberFormatException {
+    if (value == null) {
```
Comment (sappenin) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:50]: As a style, this should be: `Objects.requireNonNull(value);`

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java (hunk 2 above)
Comment (sappenin) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:71]: Yes, also see comment above. As a style, this should be: `Objects.requireNonNull(value);`

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java [hunk 4]
```diff
@@ -85,6 +109,24 @@ public LedgerIndex plus(LedgerIndex other) {
...
+   */
+  public boolean isValid() {
+    if (value.equals("current")) return true;
+    if (value.equals("validated")) return true;
+    if (value.equals("closed")) return true;
+    try {
+      unsignedLongValue();
+      return true;
+    } catch (NumberFormatException e) {
+      return false;
```
Comment (sappenin) [inline xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:126]: Is this a phantom comment, or is this actually not covered? Seems like line 31 of the unit test should trigger this. Am I missing something? Might be worth "run with coverage" in Intellij just to double-check.

**Review by sappenin — CHANGES_REQUESTED** [review]
This is really close - some very minor changes and I think we're good to go. Thanks for doing this!

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java (hunk 1 above)
Comment (sappenin) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java:48]: Yes, I agree -- it looks like your new test (see [here](https://github.com/XRPLF/xrpl4j/pull/98/files#diff-ba78bf512011908df564830de4f68373ea102c5f6ec9b3dea7a1febef3936615R31)) covers everything that was covered by `createInvalidNumericalLedgerIndex` -- feel free to remove it and verify that test coverage is the same?

**Comment by tekbar11** [comment]
I fixed the code coverage loss I introduced, jacoco now only highlights the code that was already not considered as covered before I made changes to `LexerIndex`.

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java (hunk 4 above)
Comment (nkramer44) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java]: I was confused about this as well. I can pull this down and run with coverage. Sometimes Codecov isn't 100% accurate

Comment (tekbar11) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java]: Yes this should have been considered covered, but jacoco seems to not report `throw` statements as covered.
From [the jacoco FAQ](https://www.eclemma.org/jacoco/trunk/doc/faq.html):

**Source code lines with exceptions show no coverage. Why?**

*JaCoCo determines code execution with so called probes. Probes are inserted into the control flow at certain positions. Code is considered as executed when a subsequent probe has been executed. In case of exceptions such a sequence of instructions is aborted somewhere in the middle and the corresponding line of source code is not marked as covered.*

I fixed this in my latest commit, now the code coverage should be OK.

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java (hunk 3 above)
Comment (tekbar11) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java]: done

File: xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java (hunk 2 above)
Comment (tekbar11) [reply xrpl4j-model/src/main/java/org/xrpl/xrpl4j/model/client/common/LedgerIndex.java]: done

**Review by sappenin — APPROVED** [review]
